r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+            let y_1 = self.y.get_value().get()?;
+            let x_3 = x_3.get_value().get()?;
+            Ok(lambda_val * &(x_1 - &x_3) - &y_1)
+        })?;
+
+        // Check lambda
+        lambda.mul_equals(cs.ns(|| "check lambda"), &x2_minus_x1, &y2_minus_y1)?;
+
+        // Check x3
+        let x3_plus_x1_plus_x2 = x_3
+            .add(cs.ns(|| "x3 + x1"), &self.x)?
+            .add_constant(cs.ns(|| "x3 + x1 + x2"), &other_x)?;
+        lambda.mul_equals(cs.ns(|| "check x3"), &lambda, &x3_plus_x1_plus_x2)?;
+
+        // Check y3
+        let y3_plus_y1 = y_3.add(cs.ns(|| "y3 + y1"), &self.y)?;
+        let x1_minus_x3 = self.x.sub(cs.ns(|| "x1 - x3"), &x_3)?;
+
+        lambda.mul_equals(cs.ns(|| ""), &x1_minus_x3, &y3_plus_y1)?;
+
+        Ok(Self::new(x_3, y_3, Boolean::Constant(false)))
+    }
+
+    /// Variable base exponentiation.
+    /// Inputs must be specified in *little-endian* form.
+    /// If the addition law is incomplete for the identity element,
+    /// `result` must not be the identity element.
+    fn mul_bits<'a, CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+        result: &Self,
+        bits: impl Iterator<Item = &'a Boolean>,
+    ) -> Result<Self, SynthesisError> {
+        let mut power = self.clone();
+        let mut result = result.clone();
+        for (i, bit) in bits.enumerate() {
+            let new_encoded = result.add(&mut cs.ns(|| format!("Add {}-th power", i)), &power)?;
+            result = Self::conditionally_select(
+                &mut cs.ns(|| format!("Select {}", i)),
+                bit.borrow(),
+                &new_encoded,
+                &result,
+            )?;
+            power.double_in_place(&mut cs.ns(|| format!("{}-th Doubling", i)))?;
+        }
+        Ok(result)
+    }
+
+    #[inline]
+    fn mul_bits_fixed_base<'a, CS: ConstraintSystem<ConstraintF>>(
+        base: &'a SWProjective<P>,
+        mut cs: CS,
+        result: &Self,
+        bits: &[Boolean],
+    ) -> Result<Self, SynthesisError>{
+
+        let mut to_sub = SWProjective::<P>::zero();
+
+        let mut t = base.clone();
+        let sigma = base.clone();
+        let mut result = result.clone();
+
+        let mut bit_vec = Vec::new();
+        bit_vec.extend_from_slice(bits);
+        //Simply add padding. This should be safe, since the padding bit will be part of the
+        //circuit. (It is also done elsewhere).
+        if bits.len() % 2 != 0 {
+            bit_vec.push(Boolean::constant(false))
+        }
+
+        for (i, bits) in bit_vec.chunks(2).enumerate() {
+            let ti = t.clone();
+            let two_ti = SWProjective::<P>::double(&ti);
+            let mut table = [
+                sigma,
+                sigma + &ti,
+                sigma + &two_ti,
+                sigma + &ti + &two_ti,
+            ];
+
+            //Compute constants
+            SWProjective::batch_normalization(&mut table);
```

Comment:
An even better optimization of the running time (at the cost of some extra memory) would be to do prepare the lookup table in projective arithmetics for ALL the exponent bits, and then do batch_normalization.